src/sync/sync_config.hpp

```diff
@@ -30,20 +35,57 @@ class SyncSession;
 
 enum class SyncSessionStopPolicy;
 
-enum class SyncSessionError {
-    Debug,        // An informational error, nothing to do. Only for debug purposes.
-    SessionFatal, // The session is invalid and should be killed.
-    AccessDenied, // Permissions error with the session.
-    UserFatal,    // The user associated with the session is invalid.
-};
-
 struct SyncConfig;
 using SyncBindSessionHandler = void(const std::string&,          // path on disk of the Realm file.
                                     const SyncConfig&,           // the sync configuration object.
                                     std::shared_ptr<SyncSession> // the session which should be bound.
                                     );
 
-using SyncSessionErrorHandler = void(std::shared_ptr<SyncSession>, int error_code, std::string message, SyncSessionError);
+struct SyncError;
+using SyncSessionErrorHandler = void(std::shared_ptr<SyncSession>, SyncError);
+
+struct SyncError {
+    using ProtocolError = realm::sync::ProtocolError;
+
+    std::error_code error_code;
+    std::string message;
+    bool is_fatal;
+
+    /// The error applies to the entire sync client.
+    bool is_client_error() const
+    {
+        return error_code.category() == realm::sync::client_error_category();
+    }
+
+    /// The error is a protocol or connection-level error.
+    bool is_protocol_error() const
+    {
+        if (error_code.category() != realm::sync::protocol_error_category()) {
+            return false;
+        }
+        return !realm::sync::is_session_level_error(static_cast<ProtocolError>(error_code.value()));
+    }
+
+    /// The error is specific to a session.
+    bool is_session_error() const
+    {
+        if (error_code.category() != realm::sync::protocol_error_category()) {
+            return false;
+        }
+        return !is_user_error() && realm::sync::is_session_level_error(static_cast<ProtocolError>(error_code.value()));
+    }
+
+    /// The error is specific to the user owning the session attached to this error.
+    bool is_user_error() const
+    {
+        return error_code == ProtocolError::bad_authentication;
+    }
+
+    bool is_access_denied_error() const
```

Comment:
Why bother with this getter when the the user of `SyncError` can perform this comparison itself?